File: native/bindings/endstone/permissions/permissible.cpp
#include "permissible.h"
#include <aegilex-runtime/src/cxx_host_common.rs.h>

#include "bindings/endstone/actor/actor.h"
#include "bindings/endstone/actor/mob.h"
#include "bindings/endstone/actor/player.h"
#include "bindings/endstone/command_sender.h"
#include "bindings/endstone/permissions/permission_attachment.h"
#include "bindings/endstone/server.h"

#include <endstone/command/command_sender.h>
#include <endstone/actor/actor.h>
#include <endstone/actor/mob.h>
#include <endstone/player.h>
#include <endstone/permissions/permissible.h>
#include <endstone/permissions/permission_attachment_info.h>
#include <endstone/plugin/plugin_manager.h>
#include <endstone/server.h>

#include <algorithm>
#include <string>
#include <string_view>

namespace aegilex::native::host {

namespace {

[[nodiscard]] bool valid_permission_name(const std::string_view name) noexcept
{
    if (name.empty()) {
        return false;
    }
    return std::all_of(name.begin(), name.end(), [](const char value) {
        return (value >= 'a' && value <= 'z') || (value >= '0' && value <= '9') || value == '.' || value == '_' ||
               value == '-';
    });
}

} // namespace

class Permissible::impl {
  public:
    explicit impl(endstone::Permissible *permissible) noexcept : permissible(permissible)
    {
    }

    endstone::Permissible *permissible;
};

Permissible::Permissible(endstone::Permissible *permissible) noexcept
    : impl(std::make_shared<class Permissible::impl>(permissible))
{
}

endstone::Permissible *Permissible::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->permissible;
}

PermissionLevel Permissible::getPermissionLevel() const
{
    try {
        switch (impl->permissible->getPermissionLevel()) {
        case endstone::PermissionLevel::Operator:
            return PermissionLevel::Operator;
        case endstone::PermissionLevel::Console:
            return PermissionLevel::Console;
        case endstone::PermissionLevel::Default:
            return PermissionLevel::Default;
        }
        return PermissionLevel::Default;
    }
    catch (...) {
        return PermissionLevel::Default;
    }
}

bool Permissible::isPermissionSet(const rust::Str name) const
{
    try {
        return impl->permissible->isPermissionSet(std::string(name.data(), name.size()));
    }
    catch (...) {
        return false;
    }
}

bool Permissible::hasPermission(const rust::Str name) const
{
    try {
        return impl->permissible->hasPermission(std::string(name.data(), name.size()));
    }
    catch (...) {
        return false;
    }
}

EffectivePermission Permissible::getEffectivePermission(const rust::Str name) const
{
    try {
        const std::string permission(name.data(), name.size());
        for (const auto *entry : impl->permissible->getEffectivePermissions()) {
            if (entry != nullptr && entry->getPermission() == permission) {
                return EffectivePermission{.has = true, .value = entry->getValue()};
            }
        }
    }
    catch (...) {
    }
    return EffectivePermission{.has = false, .value = false};
}

rust::Vec<rust::String> Permissible::listEffectivePermissionNames() const
{
    rust::Vec<rust::String> names;
    try {
        for (const auto *entry : impl->permissible->getEffectivePermissions()) {
            if (entry != nullptr) {
                names.push_back(rust::String(entry->getPermission()));
            }
        }
    }
    catch (...) {
    }
    return names;
}

std::unique_ptr<::aegilex::native::admin::PermissionAttachment>
Permissible::attach(const ::aegilex::native::server::Server &server, const rust::Str name, const bool value) const
{
    try {
        if (native() == nullptr || server.native() == nullptr ||
            !valid_permission_name(std::string_view(name.data(), name.size()))) {
            return {};
        }
        auto *plugin = server.native()->getPluginManager().getPlugin("aegilex");
        auto *attachment = plugin == nullptr ? nullptr : native()->addAttachment(*plugin, std::string(name), value);
        return attachment == nullptr ? std::unique_ptr<::aegilex::native::admin::PermissionAttachment>()
                                     : std::make_unique<::aegilex::native::admin::PermissionAttachment>(attachment);
    }
    catch (...) {
        return {};
    }
}

std::unique_ptr<::aegilex::native::admin::PermissionAttachment>
Permissible::attachEmpty(const ::aegilex::native::server::Server &server) const
{
    try {
        if (native() == nullptr || server.native() == nullptr) {
            return {};
        }
        auto *plugin = server.native()->getPluginManager().getPlugin("aegilex");
        auto *attachment = plugin == nullptr ? nullptr : native()->addAttachment(*plugin);
        return attachment == nullptr ? std::unique_ptr<::aegilex::native::admin::PermissionAttachment>()
                                     : std::make_unique<::aegilex::native::admin::PermissionAttachment>(attachment);
    }
    catch (...) {
        return {};
    }
}

std::unique_ptr<::aegilex::native::admin::PermissionAttachment>
Permissible::getEffectiveAttachment(const rust::Str permission) const
{
    try {
        if (native() == nullptr) {
            return {};
        }
        const std::string name(permission.data(), permission.size());
        for (const auto *entry : native()->getEffectivePermissions()) {
            if (entry != nullptr && entry->getPermission() == name && entry->getAttachment() != nullptr) {
                return std::make_unique<::aegilex::native::admin::PermissionAttachment>(entry->getAttachment());
            }
        }
    }
    catch (...) {
    }
    return {};
}

void Permissible::recalculatePermissions() const
{
    try {
        impl->permissible->recalculatePermissions();
    }
    catch (...) {
    }
}

std::unique_ptr<Permissible> permissible_from_command_sender(const CommandSender &sender) noexcept
{
    try {
        return sender.native() == nullptr
                   ? std::unique_ptr<Permissible>()
                   : std::make_unique<Permissible>(static_cast<endstone::Permissible *>(sender.native()));
    }
    catch (...) {
        return std::unique_ptr<Permissible>();
    }
}

std::unique_ptr<CommandSender> asCommandSender(const Permissible &permissible,
                                               const ::aegilex::native::server::Server &server) noexcept
{
    try {
        auto *sender = permissible.native() == nullptr ? nullptr : permissible.native()->asCommandSender();
        return sender == nullptr ? std::unique_ptr<CommandSender>()
                                 : std::make_unique<CommandSender>(sender, server.native());
    }
    catch (...) {
        return std::unique_ptr<CommandSender>();
    }
}

} // namespace aegilex::native::host

File: rust/crates/aegilex-runtime/src/host/endstone/permissions/permissible.rs
//! Core ABI implementation for `native/bindings/endstone/permissions/permissible.h`.

use crate::host::endstone::support::*;

impl crate::core_host::imports::HostPermissible for PluginStoreState {
    fn permissible_get_permission_level(
        &mut self,
        self_: u32,
    ) -> Result<Result<PermissionLevelPermissionLevel, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.get-permission-level")?;
            resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|permissible| match permissible.getPermissionLevel() {
                cxx_common::PermissionLevel::Default => PermissionLevelPermissionLevel::Default,
                cxx_common::PermissionLevel::Operator => PermissionLevelPermissionLevel::Operator,
                cxx_common::PermissionLevel::Console => PermissionLevelPermissionLevel::Console,
                _ => PermissionLevelPermissionLevel::Default,
            })
            .map_err(map_core_host_error)
        })())
    }

    fn permissible_has_permission(
        &mut self,
        self_: u32,
        permission: String,
    ) -> Result<Result<bool, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.has-permission")?;
            resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|permissible| permissible.hasPermission(&permission))
            .map_err(map_core_host_error)
        })())
    }

    fn permissible_is_permission_set(
        &mut self,
        self_: u32,
        permission: String,
    ) -> Result<Result<bool, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.is-permission-set")?;
            resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|permissible| permissible.isPermissionSet(&permission))
            .map_err(map_core_host_error)
        })())
    }

    fn permissible_recalculate_permissions(
        &mut self,
        self_: u32,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.recalculate-permissions")?;
            resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|permissible| permissible.recalculatePermissions())
            .map_err(map_core_host_error)
        })())
    }

    fn permissible_get_command_sender(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.get-command-sender")?;
            let sender = resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .and_then(|permissible| {
                resolve_server(self).map(|server| cxx_common::asCommandSender(permissible, server))
            })
            .map_err(map_core_host_error)?;
            (!sender.is_null())
                .then(|| {
                    self.insert_command_sender_resource(sender)
                        .map_err(map_core_host_error)
                })
                .transpose()
        })())
    }

    fn permissible_attach(
        &mut self,
        self_: u32,
        name: String,
        value: bool,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.attach")?;
            let attachment = resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .and_then(|permissible| {
                resolve_server(self).map(|server| permissible.attach(server, &name, value))
            })
            .map_err(map_core_host_error)?;
            self.insert_permission_attachment_resource(attachment)
                .map_err(map_core_host_error)
        })())
    }

    fn permissible_attach_empty(
        &mut self,
        self_: u32,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.attach-empty")?;
            let attachment = resolve_permissible(
                self,
                permissible_handle(self, self_).map_err(map_core_host_error)?,
            )
            .and_then(|permissible| {
                resolve_server(self).map(|server| permissible.attachEmpty(server))
            })
            .map_err(map_core_host_error)?;
            self.insert_permission_attachment_resource(attachment)
                .map_err(map_core_host_error)
        })())
    }

    fn permissible_list_effective(
        &mut self,
        self_: u32,
    ) -> Result<Result<Vec<PermissibleEffectivePermission>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.permissible.list-effective")?;
            let handle = permissible_handle(self, self_).map_err(map_core_host_error)?;
            let permissions = resolve_permissible(self, handle)
                .map(|permissible| permissible.listEffectivePermissionNames())
                .map_err(map_core_host_error)?;
            permissions
                .into_iter()
                .map(|name| {
                    let name = name.to_string();
                    let result = resolve_permissible(self, handle)
                        .map(|permissible| permissible.getEffectivePermission(&name))
                        .map_err(map_core_host_error)?;
                    let attachment = resolve_permissible(self, handle)
                        .map(|permissible| permissible.getEffectiveAttachment(&name))
                        .map_err(map_core_host_error)?;
                    let attachment = (!attachment.is_null())
                        .then(|| {
                            self.insert_permission_attachment_resource(attachment)
                                .map_err(map_core_host_error)
                        })
                        .transpose()?;
                    Ok(PermissibleEffectivePermission {
                        name,
                        value: result.value,
                        attachment,
                    })
                })
                .collect()
        })())
    }

    fn get(&mut self, sender: u32) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permissible.get")?;
            let permissible = resolve_sender(
                self,
                sender_handle(self, sender).map_err(map_core_host_error)?,
            )
            .map(cxx_common::permissible_from_command_sender)
            .map_err(map_core_host_error)?;
            self.insert_permissible_resource(permissible)
                .map_err(map_core_host_error)
        })())
    }
}
